src/main/resources/jupyter/edit-mode.js

```diff
@@ -0,0 +1,337 @@
+        const isEditMode = res.syncMode == "EDIT"
+        if (isEditMode) {
+            getLock()
+        }
+
+        renderModeBanner(isEditMode)
+    }
+
+    async function getLock() {
+        const payload = {
+            ...basePayload,
+            method: 'POST',
+            body: JSON.stringify(getLocalPath())
+        }
+
+        fetch(lockUrl, payload)
+            .then(res => {
+                handleLockStatus(res)
+                return res.json()
+            })
+            .catch(err => {
+                console.error(err)
+            })
+    }
+
+    async function toggleMetaFailureBanner(shouldShow) {
+        const bannerId = "notification_metaFailure"
+        const bannerText = "Failed to check notebook status, changes may not be saved to workspace. Retrying..."
+
+        removeElementById(bannerId)
+
+        if (shouldShow) {
+            const bannerStyling = "btn btn-xs navbar-btn btn-danger"
+
+            $('#notification_area').append(
+                $('<div>').attr({
+                    "id": bannerId,
+                    "class": bannerStyling
+                })
+                .append($('<span>').html('<i class="fa fa-exclamation-triangle"></i> ' + bannerText))
+            );
+        }
+
+    }
+
+    async function handleLockStatus(res) {
+        if (!res.ok) {
+            const status = res.status
+            const errorText = res.statusText
+
+            if (status == 409) {
+                res.json().then(body => {
+                    const message = getLockConflictBody(body.lockedBy)
+                    promptUserWithModal(lockConflictTitle, lockIssueButtons, message)
+                })
+            }
+            //for the lock endpoint, we consider all non 'ok' statuses an error
+            throw new Error(errorText)
+        }
+    }
+
+    const getLockConflictBody = (lockHolder) => {
+        return `<p>This file is currently being editted by ${lockHolder}.</p>` +
+            `<br/><p>You can make a copy, or run it in Playground Mode to explore and execute its contents without saving any changes.`;
+    }
+
+    async function promptUserWithModal(title, buttons, htmlBody) {
+        if (modalOpen) return
+
+        modalOpen = true
+
+        dialog.modal({
+                body: $('<p>').html(htmlBody),
+                title: title,
+                buttons: buttons,
+                notebook: Jupyter.notebook,
+                keyboard_manager: Jupyter.notebook.keyboard_manager
+            })
+            .on('hidden.bs.modal', () => modalOpen = false)
+            .attr('id', 'leoUserModal')
+            .find(".close").click(() => window.history.back()) //TODO: test in docker image
+    }
+
+    async function openPlaygroundMode() {
+        const url = jupyterContentsAPIUrl + Jupyter.notebook.notebook_path
+        const newPath = meta.storageLink.localSafeModeBaseDirectory.replace('notebooks/', '') + '/' + Jupyter.notebook.notebook_name
```

Comment:
"notebooks" is not guaranteed to be in this string; to form the playground mode path:

- take full relative notebook path, e.g. dir/subdir/a.ipnyb
- replace localBaseDirectory with localSafeModeBaseDirectory (should both be prefixes of the path)

Reply:
yeah, this logic was messy because the Jupyter server automatically prefixes the path to all notebooks with notebooks/ with my local config. I expected once I get it running in the docker image that this will change.